test/provisioner/provision_ibmcloud.go

```diff
@@ -5,33 +5,455 @@ package provisioner
+	subnet, _, err := IBMCloudProps.VPC.CreateSubnet(options)
+	if err != nil {
+		return err
+	}
+	IBMCloudProps.SubnetID = *subnet.ID
+	ibmcloudTrace(fmt.Sprintf("Created subnet with ID %s.\n", IBMCloudProps.SubnetID))
+
+	if len(IBMCloudProps.SubnetID) <= 0 {
+		return errors.New("SubnetID is empty, unknown error happened when create Subnet.")
+	}
+
+	return nil
+}
+
+func deleteSubnet() error {
+	if !IBMCloudProps.IsProvNewSubnet {
+		fmt.Printf("Do not delete because using existing Subnet: %s\n", IBMCloudProps.SubnetID)
+		return nil
+	}
+
+	options := &vpcv1.DeleteSubnetOptions{}
+	options.SetID(IBMCloudProps.SubnetID)
+	ibmcloudTrace(fmt.Sprintf("Deleting subnet with ID %s.\n", IBMCloudProps.SubnetID))
+	_, err := IBMCloudProps.VPC.DeleteSubnet(options)
+
+	if err != nil {
+		return err
+	}
+	ibmcloudTrace(fmt.Sprintf("Deleted subnet with ID %s.\n", IBMCloudProps.SubnetID))
+	return nil
+}
+
+func createVpcImpl() error {
+	err := createVPC()
+	if err != nil {
+		return err
+	}
+	ibmcloudTrace("waiting vpc ready before create subnet")
+	// wait vpc ready before create subnet
+	time.Sleep(60 * time.Second)
+	return createSubnet()
+}
+
+func deleteVpcImpl() error {
+	err := deleteSubnet()
+	if err != nil {
+		return err
+	}
+	return deleteVPC()
+}
+
 // IBMCloudProvisioner implements the CloudProvisioner interface for ibmcloud.
 type IBMCloudProvisioner struct {
 }
 
-func NewIBMCloudProvisioner(network string, storage string) (*IBMCloudProvisioner, error) {
+func NewIBMCloudProvisioner(properties map[string]string) (CloudProvisioner, error) {
+	if err := initProperties(properties); err != nil {
+		return nil, err
+	}
+
+	if IBMCloudProps.IsSelfManaged {
+		return &SelfManagedClusterProvisioner{}, nil
+	}
 	return &IBMCloudProvisioner{}, nil
 }
 
-func (l *IBMCloudProvisioner) CreateCluster(ctx context.Context, cfg *envconf.Config) error {
+// IBMCloudProvisioner
+
+func (p *IBMCloudProvisioner) CreateCluster(ctx context.Context, cfg *envconf.Config) error {
+	ibmcloudTrace("CreateCluster()")
+
+	clusterInfo := containerv2.ClusterCreateRequest{
+		DisablePublicServiceEndpoint: true,
+		Name:                         IBMCloudProps.ClusterName,
+		Provider:                     "vpc-gen2",
+		WorkerPools: containerv2.WorkerPoolConfig{
+			CommonWorkerPoolConfig: containerv2.CommonWorkerPoolConfig{
+				DiskEncryption: true,
+				Flavor:         IBMCloudProps.WorkerFlavor,
+				VpcID:          IBMCloudProps.VpcID,
+				WorkerCount:    IBMCloudProps.WorkerCount,
+				Zones: []containerv2.Zone{
+					{
+						ID:       IBMCloudProps.Zone,
+						SubnetID: IBMCloudProps.SubnetID,
+					},
+				},
+			},
+		},
+	}
+	target := containerv2.ClusterTargetHeader{}
+	ibmcloudTrace(fmt.Sprintf("Creating cluster %s.\n", IBMCloudProps.ClusterName))
+	_, err := IBMCloudProps.ClusterAPI.Create(clusterInfo, target)
+	if err != nil {
+		return err
+	}
+
+	// TODO, wait cluter ready, it takes long time
```

Comment:
I think this TODO already covers the suggestion that I was going to make to check for the cluster to be deployed and give the status before completing. We'll need this before we try and deploy the operator anyway.

Reply:
Let's add it in a new PR when the overlay customization PR get ready.